### ISSUE TITLE
Does not work with react-router

### ISSUE THREAD

**Issue by davidsielert** [milankinen/livereactload#43]
Doesn't do anything.. w/ global transforms on no errors .. can see script reloading in network but nothing happens..

**Comment by davidsielert** [comment]
also tried your gist code I found on google for react router w/ this and it doesn't work aswell ..    what gives ?

**Comment by milankinen** [comment]
Do you mean this gist? https://gist.github.com/milankinen/b37314362e82314c4da3

**Comment by ebrpol** [comment]
Same issue here.

Getting the following error-message in the console:

Error: Invariant Violation: Router is already running

I tried using the gist in your comment above (putting the router loading in a window.onload), but then the router loading is never called after a hot-reload.

**Comment by milankinen** [comment]
Yeah, that is unfortunate restriction when using `react-router` - router changes are not reloaded :-(

**Comment by iamdoron** [comment]
There's another workaround which is to create a router manually.

```javascript
var HashLocation = require('react-router/lib/locations/HashLocation');

var routes = (
  <Route handler={App}>
    <Route name="login" handler={Login}/>
    <Route name="logout" handler={Logout}/>
    <Route name="about" handler={About}/>
  </Route>
);

if (!window.router) {
  window.router = Router.create({
    routes: routes,
    location: HashLocation
  });

  window.router.run((Handler) =>
    React.render(<Handler/>, document.body)
  );
} else {
  window.router.replaceRoutes(routes);
}
```

It is also possible to stop the router and create a new one.

**Comment by wenbing** [comment]
met the same problem

**Comment by desy83** [comment]
same problem here. trying to get livereactload@2.1.0 running with react-router@1.x.
the workaround from @iamdoron failed because react-router did not implement the run method at 1.x.

referred to https://github.com/rackt/react-router/blob/master/CHANGES.md
i tried to refactor the workaround with:
```
var routes = (
  <Route path="/" component={Login}/>
);
ReactDOM.render(<Router>{routes}</Router>, document.getElementById('el'));
```
but still no hot reloading.

some ideas, concepts, approaches thanks

**Comment by Elyx0** [comment]
Any news on this matter on a react routing lib that isn't conflicting ? Migrating to webpack would be quite a bummer.

**Comment by milankinen** [comment]
Actually LiveReactload 2.x works exactly same like webpack so same problems remain in both bundlers. I've always done my routing from the "long woods"  so never had problems with it. 

Hmmm.. How have webpack guys dealt with this?

**Comment by andrewthornton** [comment]
Just curious if anyone has made and progress with getting this to work. All my attempts have failed. I' confident is my setups because when I ditch router, reload happens as expected.

**Comment by milankinen** [comment]
Is live reloading working with `Webpack HMR`? If someone could give some pointers how the reloading is done with `webpack`, then those would help to setup live reloading with LiveReactload as well.

**Comment by tukkajukka** [comment]
@andrewthornton I got HMR working in my project without any problems. The project is a starter template that includes React Router: https://github.com/lemanse/simple-react-app

Hope this helps!

**Comment by milankinen** [comment]
Thanks @tukkajukka! I'll take a look at your app as soon as I can.

**Comment by sassanh** [comment]
@milankinen This is what was happening for me.
I was using a component defined as an arrow function in a `Route` directive and then hot reloading wasn't working. After I saw @tukkajukka 's example I got tempted to compare my project to his sample to find out what's the issue in mine. For me the issue was using arrow function, exact same project works if I change the arrow function to a class inheriting from `React.Component` (even though it's even wrapped in a redux `connect` wrapper, it's still working).
After that I got curious maybe arrow functions don't work with livereactload generally but it's not the case, arrow functions defined in different modules in my project used in the above mentioned component (used in `Route`) get hot reloaded perfectly.
I hope it helps.

**Comment by neemzy** [comment]
According to https://github.com/reactjs/react-router/issues/2982, the react-router folks are aware of that and seem to think the "issue" is somewhat on their side (as routing components aren't pure and thus can't easily be hot-reloaded).

**Comment by milankinen** [comment]
Thanks @sassanh for the very detailed explanation! And thanks @neemzy for the link.

Arrow function components ("pure components") are not weaved by react-hot-transform because that might cause some other side effects to functionality. That's why only ES6 class and `createClass` components are working. As a general principle I'd say that if you want to support live-reloading with `react-router`, the top-level components (components that are assigned to `Route`'s `component` property) must be either ES6 class components or components created with `React.createClass`.

What comes to `connect` wrapper, the current official version (`2.2.4`) has a bug because it relies on a private member of `react-proxy` proxies, which doesn't exist in `react-proxy@1.1.8` anymore. This should be fixed in the upcoming `2.3.0` version (beta version already available, please see #118). 

@sassanh: it would be great if you could check whether your solution works with the new beta release or not (API is unchanged, you just need to install explicit version `npm install livereactload@2.3.0-rc5`).

**Comment by milankinen** [comment]
I'm closing this issue for now. I'll create another one for the `react-router` example.
